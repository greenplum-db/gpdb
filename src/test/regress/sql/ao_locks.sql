-- @Description The locks held after different operations
DROP TABLE IF EXISTS ao;

CREATE TABLE ao (a INT, b INT) WITH (appendonly=true);
INSERT INTO ao SELECT i as a, i as b FROM generate_series(1, 100) AS i;

create or replace view locktest as
select coalesce(
  case when relname like 'pg_toast%index' then 'toast index'
       when relname like 'pg_toast%' then 'toast table'
       when relname like 'pg_aoseg%' then 'aoseg table'
       when relname like 'pg_aovisimap%index' then 'aovisimap index'
       when relname like 'pg_aovisimap%' then 'aovisimap table'
       else relname end, 'dropped table'),
  mode,
  locktype,
  case when l.gp_segment_id = -1 then 'master'
       when COUNT(*) = 1 then '1 segment'
       else 'n segments' end AS node
from pg_locks l
left outer join pg_class c on (l.relation = c.oid),
pg_database d
where relation is not null
and l.database = d.oid
and (relname <> 'gp_fault_strategy' or relname IS NULL)
and d.datname = current_database()
group by l.gp_segment_id = -1, relation, relname, locktype, mode
order by 1, 3, 2;

-- Actual test begins
BEGIN;
INSERT INTO ao VALUES (200, 200);
SELECT * FROM locktest;
COMMIT;

BEGIN;
DELETE FROM ao WHERE a = 1;
SELECT * FROM locktest;
COMMIT;

BEGIN;
UPDATE ao SET b = -1 WHERE a = 2;
SELECT * FROM locktest;
COMMIT;
