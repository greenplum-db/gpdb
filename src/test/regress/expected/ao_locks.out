-- @Description The locks held after different operations
DROP TABLE IF EXISTS ao;
CREATE TABLE ao (a INT, b INT) WITH (appendonly=true);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
INSERT INTO ao SELECT i as a, i as b FROM generate_series(1, 100) AS i;
create or replace view locktest as
select coalesce(
  case when relname like 'pg_toast%index' then 'toast index'
       when relname like 'pg_toast%' then 'toast table'
       when relname like 'pg_aoseg%' then 'aoseg table'
       when relname like 'pg_aovisimap%index' then 'aovisimap index'
       when relname like 'pg_aovisimap%' then 'aovisimap table'
       else relname end, 'dropped table'),
  mode,
  locktype,
  case when l.gp_segment_id = -1 then 'master'
       when COUNT(*) = 1 then '1 segment'
       else 'n segments' end AS node
from pg_locks l
left outer join pg_class c on (l.relation = c.oid),
pg_database d
where relation is not null
and l.database = d.oid
and (relname <> 'gp_fault_strategy' or relname IS NULL)
and d.datname = current_database()
group by l.gp_segment_id = -1, relation, relname, locktype, mode
order by 1, 3, 2;
-- Actual test begins
BEGIN;
INSERT INTO ao VALUES (200, 200);
SELECT * FROM locktest;
          coalesce          |        mode         |         locktype         |   node    
----------------------------+---------------------+--------------------------+-----------
 ao                         | AccessExclusiveLock | append-only segment file | 1 segment
 ao                         | AccessExclusiveLock | append-only segment file | master
 ao                         | RowExclusiveLock    | relation                 | 1 segment
 ao                         | RowExclusiveLock    | relation                 | master
 locktest                   | AccessShareLock     | relation                 | master
 pg_class                   | AccessShareLock     | relation                 | master
 pg_class_oid_index         | AccessShareLock     | relation                 | master
 pg_class_relname_nsp_index | AccessShareLock     | relation                 | master
 pg_locks                   | AccessShareLock     | relation                 | master
(9 rows)

COMMIT;
BEGIN;
DELETE FROM ao WHERE a = 1;
SELECT * FROM locktest;
          coalesce          |        mode         |         locktype         |   node    
----------------------------+---------------------+--------------------------+-----------
 ao                         | AccessExclusiveLock | append-only segment file | master
 ao                         | ExclusiveLock       | relation                 | master
 ao                         | RowExclusiveLock    | relation                 | 1 segment
 aovisimap index            | RowExclusiveLock    | relation                 | 1 segment
 aovisimap table            | RowExclusiveLock    | relation                 | 1 segment
 locktest                   | AccessShareLock     | relation                 | master
 pg_class                   | AccessShareLock     | relation                 | master
 pg_class_oid_index         | AccessShareLock     | relation                 | master
 pg_class_relname_nsp_index | AccessShareLock     | relation                 | master
 pg_locks                   | AccessShareLock     | relation                 | master
(10 rows)

COMMIT;
BEGIN;
UPDATE ao SET b = -1 WHERE a = 2;
SELECT * FROM locktest;
          coalesce          |        mode         |         locktype         |   node    
----------------------------+---------------------+--------------------------+-----------
 ao                         | AccessExclusiveLock | append-only segment file | 1 segment
 ao                         | AccessExclusiveLock | append-only segment file | master
 ao                         | ExclusiveLock       | relation                 | master
 ao                         | RowExclusiveLock    | relation                 | 1 segment
 aovisimap index            | RowExclusiveLock    | relation                 | 1 segment
 aovisimap table            | RowExclusiveLock    | relation                 | 1 segment
 locktest                   | AccessShareLock     | relation                 | master
 pg_class                   | AccessShareLock     | relation                 | master
 pg_class_oid_index         | AccessShareLock     | relation                 | master
 pg_class_relname_nsp_index | AccessShareLock     | relation                 | master
 pg_locks                   | AccessShareLock     | relation                 | master
(11 rows)

COMMIT;
